Sources/SwiftBuildSupport/SwiftBuildSystem.swift

```diff
@@ -174,6 +174,422 @@ private final class PlanningOperationDelegate: SWBPlanningOperationDelegate, Sen
+                    if let taskSignature = key.taskSignature {
+                        self.taskSignatureBuffer[taskSignature] = newValue
+                    } else if let targetID = key.targetID {
+                        self.targetIDBuffer[targetID] = newValue
+                    }
+                }
+            }
+
+            subscript(task: SwiftBuildMessage.TaskStartedInfo) -> Data? {
+                get {
+                    guard let result = self.taskSignatureBuffer[task.taskSignature] else {
+                        // Default to checking targetID and taskID.
+                        if let result = self.taskIDBuffer[task.taskID] {
+                            return result
+                        } else if let targetID = task.targetID,
+                                  let result = self.targetIDBuffer[targetID] {
+                            return result
+                        }
+
+                        // Return global buffer if none of the above are found.
+                        return self.globalBuffer
+                    }
+
+                    return result
+                }
+            }
+        }
+
+        mutating func started(task: SwiftBuild.SwiftBuildMessage.TaskStartedInfo) throws {
+            if activeTasks[task.taskID] != nil {
+                throw Diagnostics.fatalError
+            }
+            activeTasks[task.taskID] = task
+            taskIDToSignature[task.taskID] = task.taskSignature
+        }
+
+        mutating func completed(task: SwiftBuild.SwiftBuildMessage.TaskCompleteInfo) throws -> SwiftBuild.SwiftBuildMessage.TaskStartedInfo {
+            guard let startedTaskInfo = activeTasks[task.taskID] else {
+                throw Diagnostics.fatalError
+            }
+            if completedTasks[task.taskID] != nil {
+                throw Diagnostics.fatalError
+            }
+            self.completedTasks[task.taskID] = task
+            return startedTaskInfo
+        }
+
+        mutating func started(target: SwiftBuild.SwiftBuildMessage.TargetStartedInfo) throws {
+            if targetsByID[target.targetID] != nil {
+                throw Diagnostics.fatalError
+            }
+            targetsByID[target.targetID] = target
+        }
+
+        func target(for task: SwiftBuild.SwiftBuildMessage.TaskStartedInfo) throws -> SwiftBuild.SwiftBuildMessage.TargetStartedInfo? {
+            guard let id = task.targetID else {
+                return nil
+            }
+            guard let target = targetsByID[id] else {
+                throw Diagnostics.fatalError
+            }
+            return target
+        }
+
+        func taskSignature(for id: Int) -> String? {
+            if let signature = taskIDToSignature[id] {
+                return signature
+            }
+            return nil
+        }
+
+        mutating func appendToBuffer(_ info: SwiftBuildMessage.OutputInfo) {
+            // Attempt to key by taskSignature; at times this may not be possible,
```

Comment:
The reason this taskID vs taskSignature situation exists was because we started to transition from IDs (which change on every build operation) to signatures (which are stable across build operations), but didn't complete it, and also ran into performance issues (106579386) and turned them off for console output at least... I don't remember why and maybe we can re-enable them at some point, but we should definitely clean this up (not that it should block this PR though).